### DISCUSSION TITLE
How to make child components reactive

### DISCUSSION THREAD

**Discussion by marcotas** [livewire/livewire#2097] [2. Show and tell]
Hello everyone!

I just would like to share with the community the simple approach I use to make a livewire child component "reactive" based on parent changes. In your parent livewire component you use a timestamp for the key attribute:
```php
<livewire:child-component key="{{ now() }}" :my-prop="$myModel" />
```
The tricky part is the `now()` function in `key` prop.
This component now will always be updated when the parent is updated. What do you think about this?
I hope that could help someone else.

### Important Update
We need to use a unique value for `key` attribute used by livewire components. This way we will avoid this [kind of error](https://github.com/livewire/livewire/issues/1961#issuecomment-721567668) that I faced. One alternative would be using `Str::random()` instead of `now()` like this:
```php
<livewire:child-component key="{{ Str::random() }}" :my-prop="$myModel" />
```
Or adding a prefix or suffix to it: `'child-component-' . now() . $myModel->id`.

**Comment by mblarsen** [comment]
Your approach is fine if you don't mind everything updating all the time.

I try to use meaningful keys whenever I can. If you are unable to construct that represents the state of your child component (E.g. if you used the `$myModel->id` and something outside of the child component updates $myModel then the child component wont update; the fix here would be to include that piece of data in the key along with the id), then you send events to the child component to update one or all children. E.g. `myModelUpdated` or simply `$refresh` and then add these as listeners on the child model component. If you know which child component represents the data that is refreshed e.g. $myModel with id 42. The you can use dynamic listeners e.g. 

```php
protected function getListeners()
{
        return ['myModelUpdate'. $myModel->id  => 'update'];
}
```

That way you can target a specific child. A bit extra work, but useful if you have long lists.

**Reply by marcotas** [reply]
Thank you for sharing this! That's definitely useful for updating specific items on lists. 👏

**Comment by Yinci** [comment]
This works, but I can't wrap my head around _why_ it works. How come that by using a random string or timestamp as key, it forces the child component to be responsive? When just using a model id, it simply isn't, and I don't understand why.

Why is `:wire:key="$model->id . now()"` responsive, but `:wire:key="$model->id"` isn't?

**Reply by joshhanley** [reply]
@Yinci technically it's a hack. You will find if you put time or random string in the key, that the child component will be replaced on every render of the parent component. So if you had any state in the child component that would actually be wiped out and replaced.

The reason you need it, if you want the child component to be reactive, is LW's components run independent of each other once they are loaded onto the page. It's intended that if you want to keep data in sync between components, that you fire events between them.

But that's not ideal, hence why people have found this hack of making LW replace the child component on every render by changing the key each time.

Hope this helps!

**Reply by Yinci** [reply]
>  by changing the key each time.

Aah! Now the lightbulb's screwed in! I hadn't realized that the key is regenerated each time. That makes a ton of sense now!

**Comment by angelokezimana** [comment]
Thank you sharing this amazing trick @marcotas

**Comment by alexiovay** [comment]
This should actually be in the documentation of Livewire. Also it is worth to mention that your child variable won't update, when you assign a value to this variable with the `mount` function.

**Reply by joshhanley** [reply]
@AlexioVay as I mentioned above, this is a bit of a hack to get it running, so I don't think putting it the docs is a great idea. The concern would be people using it for everything, which has significant performance implications on the front end if any nested components are getting rebuilt on every request due to the key changing each time (because of `now()`).

In terms of reactivity, we do already have this:

> Nested components CAN accept data parameters from their parents, HOWEVER they are not reactive like props from a Vue component.

Is that what you were thinking or something else?

Hope this helps!

**Comment by cfuent-es** [comment]
Hello,

how would I go about making the child component reactive without updating the parent component? 

I have a document search that enables a modal via Javascript, but I render the modal component already in the parent component's rendering, and I'm trying to update via $this->emit event, passing the Model found, but it doesn't do the data binding.

**Comment by yungifez** [comment]
I've been having issues using this where the child component multiplies for some reason
